src/main/java/uk/ac/cam/cl/dtg/segue/etl/ContentIndexer.java

```diff
@@ -629,42 +636,63 @@ private synchronized void buildElasticSearchIndex(final String sha,
             }
         }
 
+        long startTime, endTime;
 
         try {
             es.indexObject(sha, CONTENT_INDEX_TYPE.METADATA.toString(),
                     objectMapper.writeValueAsString(ImmutableMap.of("version", sha, "created", new Date().toString())), "general");
             es.indexObject(sha, CONTENT_INDEX_TYPE.METADATA.toString(),
                     objectMapper.writeValueAsString(ImmutableMap.of("tags", tagsList)), "tags");
 
-            // TODO: Should probably bulk index these
-            for (String k : allUnits.keySet()) {
-                es.indexObject(sha, CONTENT_INDEX_TYPE.UNIT.toString(),
-                        objectMapper.writeValueAsString(ImmutableMap.of("cleanKey", k, "unit", allUnits.get(k))));
-            }
-            for (String k : publishedUnits.keySet()) {
-                es.indexObject(sha, CONTENT_INDEX_TYPE.PUBLISHED_UNIT.toString(),
-                        objectMapper.writeValueAsString(ImmutableMap.of("cleanKey", k, "unit", publishedUnits.get(k))));
-            }
+            startTime = System.nanoTime();
+            es.bulkIndex(sha, CONTENT_INDEX_TYPE.UNIT.toString(), allUnits.entrySet().stream().map(entry -> {
+                try {
+                    return immutableEntry((String) null, objectMapper.writeValueAsString(ImmutableMap.of("cleanKey", entry.getKey(), "unit", entry.getValue())));
```

Comment:
It's a shame to have to create an immutableEntry with null as the first value for these.
It would be nice to be able to override the `bulkIndex` method, one accepting a List of Strings and one accepting the current List of Entries. Unfortunately, we can't overload it as Java erases the type parameter in generics and so would not be able to decide which of the two methods to choose from during runtime. However it would be fine if one had a different name so I would suggest renaming the current one to something like `bulkIndexWithIDs(...)` and then creating a `bulkIndex(...)` which hands off the actual work to `bulkIndexWithIDs(...)` after transforming the `dataToIndex` using a stream like you have here.

Reply:
Agreed, I was trying to get an MVP working and felt that it might be okay to leave it as is, but it would much cleaner and clearer to have a `bulkIndexWithIDs(...)` as you suggest. Will get that working 👍